File: backend/app/schemas/shopping.py
from pydantic import BaseModel
from uuid import UUID
from datetime import date, datetime
from typing import Optional
from decimal import Decimal


# --- Shop ---
class ShopCreate(BaseModel):
    name: str
    address: Optional[str] = None
    latitude: Optional[Decimal] = None
    longitude: Optional[Decimal] = None
    phone: Optional[str] = None
    category: Optional[str] = None
    is_favorite: bool = False
    notes: Optional[str] = None


class ShopUpdate(BaseModel):
    name: Optional[str] = None
    address: Optional[str] = None
    latitude: Optional[Decimal] = None
    longitude: Optional[Decimal] = None
    phone: Optional[str] = None
    category: Optional[str] = None
    is_favorite: Optional[bool] = None
    notes: Optional[str] = None


class ShopResponse(BaseModel):
    id: UUID
    user_id: UUID
    name: str
    address: Optional[str] = None
    latitude: Optional[Decimal] = None
    longitude: Optional[Decimal] = None
    phone: Optional[str] = None
    category: Optional[str] = None
    is_favorite: bool
    notes: Optional[str] = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


# --- Purchase Item ---
class PurchaseItemCreate(BaseModel):
    item_name: str
    category: Optional[str] = None
    quantity: Optional[Decimal] = None
    unit: Optional[str] = None
    mrp: Optional[Decimal] = None
    discount: Decimal = Decimal("0")
    bought_price: Decimal


class PurchaseItemResponse(BaseModel):
    id: UUID
    bill_id: UUID
    item_name: str
    category: Optional[str] = None
    quantity: Optional[Decimal] = None
    unit: Optional[str] = None
    mrp: Optional[Decimal] = None
    discount: Optional[Decimal] = None
    bought_price: Decimal
    created_at: datetime

    model_config = {"from_attributes": True}


# --- Shopping Bill ---
class BillCreate(BaseModel):
    shop_id: Optional[UUID] = None
    bill_number: Optional[str] = None
    bill_date: date
    total_amount: Decimal
    discount_amount: Decimal = Decimal("0")
    tax_amount: Decimal = Decimal("0")
    payment_method: Optional[str] = None
    image_url: Optional[str] = None
    entry_method: str = "manual"
    notes: Optional[str] = None
    items: list[PurchaseItemCreate] = []


class BillUpdate(BaseModel):
    shop_id: Optional[UUID] = None
    bill_number: Optional[str] = None
    bill_date: Optional[date] = None
    total_amount: Optional[Decimal] = None
    discount_amount: Optional[Decimal] = None
    tax_amount: Optional[Decimal] = None
    payment_method: Optional[str] = None
    image_url: Optional[str] = None
    notes: Optional[str] = None


class BillResponse(BaseModel):
    id: UUID
    user_id: UUID
    shop_id: Optional[UUID] = None
    bill_number: Optional[str] = None
    bill_date: date
    total_amount: Decimal
    discount_amount: Optional[Decimal] = None
    tax_amount: Optional[Decimal] = None
    payment_method: Optional[str] = None
    image_url: Optional[str] = None
    entry_method: str
    notes: Optional[str] = None
    items: list[PurchaseItemResponse] = []
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


# --- Checklist ---
class ChecklistItemCreate(BaseModel):
    item_name: str
    quantity: Optional[str] = None
    sort_order: int = 0


class ChecklistItemResponse(BaseModel):
    id: UUID
    checklist_id: UUID
    item_name: str
    quantity: Optional[str] = None
    is_checked: bool
    sort_order: int
    created_at: datetime

    model_config = {"from_attributes": True}


class ChecklistCreate(BaseModel):
    title: str
    items: list[ChecklistItemCreate] = []


class ChecklistUpdate(BaseModel):
    title: Optional[str] = None
    is_active: Optional[bool] = None


class ChecklistResponse(BaseModel):
    id: UUID
    user_id: UUID
    title: str
    is_active: bool
    items: list[ChecklistItemResponse] = []
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


# --- Analytics ---
class MonthlySpending(BaseModel):
    month: str  # "2024-01"
    total: Decimal


class ShopSpending(BaseModel):
    shop_id: UUID
    shop_name: str
    total: Decimal
    bill_count: int


class ItemFrequency(BaseModel):
    item_name: str
    count: int
    avg_price: Decimal


class ShoppingAnalytics(BaseModel):
    monthly_spending: list[MonthlySpending] = []
    shop_spending: list[ShopSpending] = []
    top_items: list[ItemFrequency] = []
    total_this_month: Decimal = Decimal("0")
    total_last_month: Decimal = Decimal("0")

File: mobile/src/types/shopping.ts
export interface Shop {
  id: string;
  user_id: string;
  name: string;
  address: string | null;
  latitude: number | null;
  longitude: number | null;
  phone: string | null;
  gstin: string | null;
  category: string | null;
  is_favorite: boolean;
  notes: string | null;
  created_at: string;
  updated_at: string;
}

export interface RecentShop {
  id: string;
  name: string;
  address: string | null;
  phone: string | null;
  gstin: string | null;
  last_visit_date: string | null;
  bill_count: number;
  total_spent: number;
}

export interface ShopCreate {
  name: string;
  address?: string | null;
  latitude?: number | null;
  longitude?: number | null;
  phone?: string | null;
  gstin?: string | null;
  category?: string | null;
  is_favorite?: boolean;
  notes?: string | null;
}

export interface PurchaseItem {
  id: string;
  bill_id: string;
  item_name: string;
  brand_name: string | null;
  category: string | null;
  quantity: number | null;
  unit: string | null;
  mrp: number | null;
  discount: number | null;
  bought_price: number;
  created_at: string;
}

export interface PurchaseItemCreate {
  item_name: string;
  brand_name?: string | null;
  category?: string | null;
  quantity?: number | null;
  unit?: string | null;
  mrp?: number | null;
  discount?: number;
  bought_price: number;
}

export interface ShoppingBill {
  id: string;
  user_id: string;
  shop_id: string | null;
  bill_number: string | null;
  bill_date: string;
  total_amount: number;
  discount_amount: number | null;
  tax_amount: number | null;
  payment_method: string | null;
  purchase_mode: string;
  image_url: string | null;
  entry_method: string;
  notes: string | null;
  items: PurchaseItem[];
  created_at: string;
  updated_at: string;
}

export interface BillCreate {
  shop_id?: string | null;
  bill_number?: string | null;
  bill_date: string;
  total_amount: number;
  discount_amount?: number;
  tax_amount?: number;
  payment_method?: string | null;
  purchase_mode?: string;
  image_url?: string | null;
  entry_method?: string;
  notes?: string | null;
  items: PurchaseItemCreate[];
}

export interface BillUpdate {
  shop_id?: string | null;
  bill_number?: string | null;
  bill_date?: string;
  total_amount?: number;
  discount_amount?: number;
  tax_amount?: number;
  payment_method?: string | null;
  purchase_mode?: string;
  image_url?: string | null;
  notes?: string | null;
  items?: PurchaseItemCreate[];
}

export interface ShoppingChecklist {
  id: string;
  user_id: string;
  title: string;
  is_active: boolean;
  items: ChecklistItem[];
  created_at: string;
  updated_at: string;
}

export interface ChecklistItem {
  id: string;
  checklist_id: string;
  item_name: string;
  quantity: string | null;
  is_checked: boolean;
  sort_order: number;
  created_at: string;
}

export interface ShoppingAnalytics {
  monthly_spending: { month: string; total: number }[];
  shop_spending: { shop_id: string; shop_name: string; total: number; bill_count: number }[];
  top_items: { item_name: string; count: number; avg_price: number }[];
  total_this_month: number;
  total_last_month: number;
}

export interface ItemPriceComparison {
  shop_name: string;
  shop_id: string;
  min_price: number;
  max_price: number;
  avg_price: number;
  last_bought_date: string;
  last_price: number;
}

// OCR extraction result (before saving)
export interface OCRBillItem {
  item_name: string;
  brand_name?: string | null;
  quantity: number | null;
  unit: string | null;
  mrp: number | null;
  discount: number | null;
  bought_price: number;
}

export interface OCRBillResult {
  shop_name: string | null;
  shop_address: string | null;
  shop_phone: string | null;
  shop_gstin: string | null;
  bill_date: string | null;
  items: OCRBillItem[];
  total_amount: number | null;
  raw_text: string | null;
}

export const SHOP_CATEGORIES = [
  'Grocery', 'Vegetables', 'Fruits', 'Dairy', 'Medical', 'Electronics',
  'Clothing', 'Hardware', 'Stationery', 'Other',
] as const;

export const SHOPPING_UNITS = ['kg', 'g', 'L', 'pcs', 'pack'] as const;

export const PURCHASE_MODES = ['offline', 'online'] as const;

// --- Bill Draft ---
export const BILL_DRAFT_VERSION = 1;

export interface BillDraft {
  id: string;
  entry_method: string;
  draft_data: Record<string, unknown>;
  created_at: string;
  updated_at: string;
}

export interface BillDraftCreate {
  entry_method?: string;
  draft_data: Record<string, unknown>;
}

export interface BillDraftUpdate {
  draft_data: Record<string, unknown>;
}
